### PR TITLE
Feature/27 goto definition

### PR DESCRIPTION
**Pull request by teggotic**

### CODE REVIEW

File: spec/compiler/serve_definition_spec.cr [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      :new
+        example = A.new
+    SOURCE
+
+    ctx = Mare::Compiler.compile([source], :serve_definition)
+
+    pos = ctx.serve_definition[Mare::Source::Pos.point(source, 4, 14)].not_nil!
+    pos.row.should eq 0
+    pos.col.should eq 7
+    pos.size.should eq "T".bytesize
```
Comment (jemc) [inline spec/compiler/serve_definition_spec.cr:16]: Not strictly needed, but it makes the test case make more sense:

```suggestion
    pos.size.should eq "A".bytesize
```

Comment (jemc) [inline spec/compiler/serve_definition_spec.cr:16]: Not strictly needed, but it makes the test case make more sense:

```suggestion
    pos.size.should eq "A".bytesize
```

File: src/mare/compiler.cr [hunk 2]
```diff
@@ -1,31 +1,33 @@
...
+    when :namespace        then ctx.run_whole_program(ctx.namespace)
+    when :macros           then ctx.run_copy_on_mutate(Macros)
+    when :sugar            then ctx.run_copy_on_mutate(Sugar)
+    when :refer_type       then ctx.run(ctx.refer_type)
+    when :populate         then ctx.run_copy_on_mutate(Populate)
+    when :lambda           then ctx.run_copy_on_mutate(Lambda)
+    when :refer            then ctx.run(ctx.refer)
+    when :classify         then ctx.run(ctx.classify)
+    when :jumps            then ctx.run(ctx.jumps)
     when :consumes       then ctx.run(ctx.consumes)
```
Comment (jemc) [inline src/mare/compiler.cr:14]:
```suggestion
    when :consumes         then ctx.run(ctx.consumes)
```

File: src/mare/compiler/serve_definition.cr [hunk 3]
```diff
@@ -0,0 +1,133 @@
+require "llvm"
+require "debug"
```
Comment (jemc) [inline src/mare/compiler/serve_definition.cr:2]: Note that this being here breaks the CI process. Here's the error:

```
in src/mare/compiler/serve_definition.cr:2: while requiring "debug": can't find file 'debug'
```

So I think we must remove it.
```suggestion
```

File: src/mare/server.cr [hunk 4]
```diff
@@ -85,6 +103,106 @@ class Mare::Server
...
+          dest_path = pair[1]
+
+          if user_filepath.starts_with?(dest_path)
+            user_filepath = user_filepath.sub(dest_path, host_path)
+          end
+        end
+      end
+
+      @wire.respond msg do |msg|
+        msg.result = LSP::Data::Location.new(URI.new(path: user_filepath), LSP::Data::Range.new(LSP::Data::Position.new(definition_pos.row.to_i64, definition_pos.col.to_i64), LSP::Data::Position.new(definition_pos.row.to_i64, definition_pos.col.to_i64 + (definition_pos.finish - definition_pos.start))))
```
Comment (jemc) [inline src/mare/server.cr:198]: Small formatting suggestion to break this large expression up over multiple lines:

```suggestion
        msg.result = LSP::Data::Location.new(
          URI.new(path: user_filepath),
          LSP::Data::Range.new(
            LSP::Data::Position.new(
              definition_pos.row.to_i64,
              definition_pos.col.to_i64,
            ),
            LSP::Data::Position.new(
              definition_pos.row.to_i64,
              definition_pos.col.to_i64 + (definition_pos.finish - definition_pos.start),
            ),
          ),
        )
```

**Comment by jemc** [comment]
Looks like there's two more tests failing in the `ServeHover` tests in CI:

```
  1) Mare::Compiler::ServeHover describes a local variable and its method
     Failure/Error: pos.col.should eq 4

       Expected: 4
            got: 12

     # spec/compiler/serve_hover_spec.cr:32

  2) Mare::Compiler::ServeHover describes a self-call
     Failure/Error: pos.col.should eq 4

       Expected: 4
            got: 5

     # spec/compiler/serve_hover_spec.cr:52
```

Did the changes in this PR affect the existing `ServeHover` pass?

**Comment by teggotic** [comment]
They shouldn't. Seems like it's now hovering now whole expression, but just the function name in hash example.

**Comment by jemc** [comment]
It was probably because of the change I helped you make in `Infer` where it is now assigning the analysis to both the entire call expression as well as the function name. I think it is okay.

We just need to update the tests to use the new column numbers.

**Comment by teggotic** [comment]
> It was probably because of the change I helped you make in `Infer` where it is now assigning the analysis to both the entire call expression as well as the function name. I think it is okay.
> 
> We just need to update the tests to use the new column numbers.

It looks like we don't get AST::Relate here https://github.com/jemc/mare/blob/4e6e9e1e1212bf937df2deacd71f78385fe23391/src/mare/compiler/serve_hover.cr#L126

**Pull request merged**